### PR TITLE
NOISSUE: refactor validation, code unification, more checks

### PR DESCRIPTION
**Pull request by bigbes**
subj

### CODE REVIEW

File: ledger-core/insolar/payload/validation_vstatereport.go [hunk 1]
```diff
@@ -6,51 +6,77 @@
...
-	case !object.GetLocal().Pulse().IsTimePulse():
-		fallthrough
-	case object.GetLocal().Pulse() >= currPulse:
-		return throw.New("Object pulse should be valid time pulse less that current pulse")
+	objectPulse, err := validSelfScopedGlobalWithPulseBeforeOrEq(m.Object, currentPulse, "Object")
+	if err != nil {
+		return err
+	}
+
+	validateStatus := map[VStateReport_StateStatus]vStateReportValidateStatusFunc{
```
Comment (dgovorukhin) [inline ledger-core/insolar/payload/validation_vstatereport.go:31]: @bigbes why is it better to use map instead switch/case?

Comment (bigbes) [reply ledger-core/insolar/payload/validation_vstatereport.go:31]: Unification of check signatures, lesser code duplication; no other practical reason
switch-case may be still faster, can be rewritten as `func(VStateReport_StateStatus) vStateReportValidateStatusFunc` with switch case for functions (combining that two decisions)

Comment (dgovorukhin) [reply ledger-core/insolar/payload/validation_vstatereport.go:31]: What is the duplication code? We had 4 methods and switch/case. now we have an additional type for function + map and the same 4 methods.

Comment (bigbes) [reply ledger-core/insolar/payload/validation_vstatereport.go:31]: every `case` have exactly same code with only one difference: callback

File: ledger-core/insolar/payload/validation.go [hunk 2]
```diff
@@ -8,32 +8,200 @@ package payload
...
 }
 
-func isTimePulseBefore(pn pulse.Number, before pulse.Number) bool {
-	return pn.IsTimePulse() && pn.IsBefore(before)
+type PulseValidator interface {
+	ValidateTime(pn pulse.Number, before pulse.Number) bool
+	Relation() string
+}
+
+type ExtendedPulseValidator interface {
```
Comment (dgovorukhin) [inline ledger-core/insolar/payload/validation.go:23]: @bigbes If this interface named "extended" PulseValidator maybe we can inherit it?
Signature methods is matched
```
type ExtendedPulseValidator interface {
	PulseValidator
	GetSpecialValidator() PulseValidator
	ValidateSpecialOrTime(pn pulse.Number, before pulse.Number) bool
}
```

File: ledger-core/insolar/payload/validation_vstatereport.go (hunk 1 above)
Comment (dgovorukhin) [reply ledger-core/insolar/payload/validation_vstatereport.go:31]: This code example the same simple, but no need a new type and map.
```
var err error

	switch m.GetStatus() {
	case Ready:
		err = m.validateStatusReady(objectPulse, currentPulse)
	case Empty:
		err = m.validateStatusEmpty(objectPulse, currentPulse)
	case Missing, Inactive:
		err = m.validateStatusMissingOrInactive()
	default:
		return throw.New("Unexpected state received")
	}

	if err != nil {
		return err
	}
```

File: ledger-core/insolar/payload/validation.go (hunk 2 above)
Comment (bigbes) [reply ledger-core/insolar/payload/validation.go:23]: fixed

File: ledger-core/insolar/payload/validation_vstatereport.go (hunk 1 above)
Comment (bigbes) [reply ledger-core/insolar/payload/validation_vstatereport.go]: fixed

**Pull request merged**